[SPARK-7120][SPARK-7121] Closure cleaner nesting + documentation + tests

andrewor14 opened the pull request:
Note: ~600 lines of this is test code, and ~100 lines documentation.

**[SPARK-7121]** ClosureCleaner does not handle nested closures properly. For instance, in SparkContext, I tried to do the following:

```
def scope[T](body: => T): T = body // no-op
def myCoolMethod(path: String): RDD[String] = scope {
  parallelize(1 to 10).map { _ => path }
}
```

and I got an exception complaining that SparkContext is not serializable. The issue here is that the inner closure is getting its path from the outer closure (the scope), but the outer closure references the SparkContext object itself to get the `parallelize` method.

Note, however, that the inner closure doesn't actually need the SparkContext; it just needs a field from the outer closure. If we modify ClosureCleaner to clean the outer closure recursively using only the fields accessed by the inner closure, then we can serialize the inner closure.

**[SPARK-7120]** Also, the other thing is that this file is one of the least understood, partly because it is very low level and is written a long time ago. This patch attempts to change that by adding the missing documentation.

This is blocking my effort on a separate task #5729.

Commit:
Implement transitive cleaning + add missing documentation

See in-code comments for more detail on what this means.

Commit:
Feature flag this new behavior

... in case anything breaks, we should be able to resort to old
behavior.

SparkQA commented:
[Test build #30930 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30930/consoleFull) for   PR 5685 at commit [`2390a60`](https://github.com/apache/spark/commit/2390a608ed74a9703d3763d040421dccb51242ec).

SparkQA commented:
[Test build #30930 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30930/consoleFull) for   PR 5685 at commit [`2390a60`](https://github.com/apache/spark/commit/2390a608ed74a9703d3763d040421dccb51242ec).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `   *   class SomethingNotSerializable `
  * `      logDebug(s" + cloning the object $obj of class $`
  * `class FieldAccessFinder(`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30930/
Test FAILed.

srowen commented:
You know a lot more about this than I, but I was under the impression that the closure cleaner couldn't clean beyond a level or so because it would then be modifying local object state by nulling fields in them and that's not necessarily permissible. I'm sure you're on top of that, just noting my recollection from similar discussions.

pwendell commented:
@srowen wondering, what do you mean by "local state"? The closure cleaner only nulls out fields in _clones_ of local objects (actually the nulling mechanism is really just that it creates a clone and then selectively copies over certain fields from the original and neglects others). So I think it should be fine.

aarondav commented on core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala:
```diff
@@ -167,15 +294,17 @@ private[spark] object ClosureCleaner extends Logging {
       val params = cons.getParameterTypes.map(createNullValue).toArray
-      if (outer != null) {
-        params(0) = outer // First param is always outer object
+      if (enclosingObject!= null) {
```
nit spcae

andrewor14 replied:
oops thanks

srowen commented:
Yeah I'm probably misunderstanding prior discussions, but I had the impression there was a reason you couldn't always clear references two hops away. This would be great to improve if it can be done.

andrewor14 commented:
@srowen we only cannot null fields in parent closures in place, but the closure cleaner actually clones these parent closures so it's safe to null things out there.

Commit:
Minor changes

aarondav commented on core/src/test/scala/org/apache/spark/util/ClosureCleanerSuite.scala:
```diff
@@ -50,7 +50,7 @@ class ClosureCleanerSuite extends FunSuite {
     assert(obj.run() === 96) // 4 * (1+2+3+4) + 4 * (1+2+3+4) + 16 * 1
   }
-  
+
```
The suite should be expanded with the fix, right? Could you also test the common `for` loop case that others have seen in the past?

andrewor14 replied:
coming soon

pwendell commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1740,7 +1740,8 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    */
   private[spark] def clean[F <: AnyRef](f: F, checkSerializable: Boolean = true): F = {
-    ClosureCleaner.clean(f, checkSerializable)
+    val cleanTransitively = conf.getBoolean("spark.closureCleaner.transitive", true)
```
I'm not sure that it makes sense to flag this since (provided we add scopes) then every user program will break if this is toggled. I think this feature will need to be removed wholesale if there are issues.

pwendell commented on core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala:
```diff
@@ -77,6 +80,9 @@ private[spark] object ClosureCleaner extends Logging {
+  /**
+   * Return a list of classes that represent closures enclosed in the given closure object.
+   */
   private def getInnerClasses(obj: AnyRef): List[Class[_]] = {
```
Can we do validations here that `obj` is a closure type (for instance doing an instanceOf check with an exception)? It would add a bit more type safety and clarity to the API.

andrewor14 replied:
how can I do that?

pwendell replied:
what about doing `isClosure` on the class name of the object?

Commit:
Add tests (still WIP)

The existing ones are not passing yet because cleaning closures
is not idempotent. This will be added in a future commit.

Commit:
Make closure cleaning idempotent

We need this for tests because we clean the same closure many
times there. Outside of tests this is probably not important.

Commit:
Finalize tests

mateiz commented:
Does this happen in user programs or just in SparkContext? This is exactly what ClosureCleaner was designed to deal with, so I'm surprised that it's a problem.

mateiz commented:
BTW the general workaround for this kind of stuff is to grab the values from outside the closure into a local val before you call something. I'd look into how to do that before changing the ClosureCleaner too heavily.

andrewor14 commented:
Yes, this also affects user programs. For example I modified SparkPi to follow the pattern I described in the PR description
```
    ...
    val slices = if (args.length > 0) args(0).toInt else 2
    (1 to 1).foreach { j =>
      val count = spark.parallelize(1 until n, slices).map { i =>
        val z = slices // *** This is the culprit ***
        val x = random * 2 - 1
        val y = random * 2 - 1
        if (x * x + y * y < 1) 1 else 0
      }.reduce(_ + _)
    }
    ...
```
If you run this in master branch you will run into a task not serializable exception. I was able to verify that this patch fixes this as long as `spark.closureCleaner.transitive` is enabled. If this is not enabled, if fails with the same exception as before.

andrewor14 commented:
The main issue with grabbing the values into the closure is that we'll need to do this everywhere. My intention is to wrap many existing methods (in SparkContext and other files) in closures, and it will be very difficult and manual to have to localize the variables in all of these places.

Commit:
Merge branch 'master' of github.com:apache/spark into closure-cleaner

Commit:
Fix closure cleaner visibility

Commit:
Guard against potential infinite cycles in method visitor

Now we keep track of the methods that we visited to avoid visiting
the same method twice.

Commit:
Add more tests for individual closure cleaner operations

Commit:
Use private method tester for a few things

andrewor14 commented:
Jenkins, test this please?

Commit:
Update a few comments

SparkQA commented:
[Test build #700 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/700/consoleFull) for   PR 5685 at commit [`eb127e5`](https://github.com/apache/spark/commit/eb127e54fa04ef17523806067d074b8560ea783e).

pwendell commented:
@andrewor14 I manually triggered this using Josh's tool - looks like PRB is being tempermental.

Commit:
More minor updates (wording, renaming etc.)

Commit:
Merge branch 'master' of github.com:apache/spark into closure-cleaner

SparkQA commented:
[Test build #708 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/708/consoleFull) for   PR 5685 at commit [`8b71cdb`](https://github.com/apache/spark/commit/8b71cdb7953ce622fd94fda7e0c5daafeb145cca).

SparkQA commented:
[Test build #708 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/708/consoleFull) for   PR 5685 at commit [`8b71cdb`](https://github.com/apache/spark/commit/8b71cdb7953ce622fd94fda7e0c5daafeb145cca).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch **adds the following new dependencies:**
   * `tachyon-0.6.4.jar`
   * `tachyon-client-0.6.4.jar`

 * This patch **removes the following dependencies:**
   * `tachyon-0.5.0.jar`
   * `tachyon-client-0.5.0.jar`

Commit:
Fix scala style?

SparkQA commented:
[Test build #709 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/709/consoleFull) for   PR 5685 at commit [`4aab379`](https://github.com/apache/spark/commit/4aab379c0d6bc12a9e0c4a984d87bbfc21bd948b).

SparkQA commented:
[Test build #711 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/711/consoleFull) for   PR 5685 at commit [`6d4d3f1`](https://github.com/apache/spark/commit/6d4d3f1ac8da883fb814613afec35900b078b751).

andrewor14 commented:
Weird, tests pass locally. Investigating.

SparkQA commented:
[Test build #30977 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30977/consoleFull) for   PR 5685 at commit [`6d4d3f1`](https://github.com/apache/spark/commit/6d4d3f1ac8da883fb814613afec35900b078b751).

SparkQA commented:
[Test build #30977 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30977/consoleFull) for   PR 5685 at commit [`6d4d3f1`](https://github.com/apache/spark/commit/6d4d3f1ac8da883fb814613afec35900b078b751).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch **adds the following new dependencies:**
   * `tachyon-0.6.4.jar`
   * `tachyon-client-0.6.4.jar`

 * This patch **removes the following dependencies:**
   * `tachyon-0.5.0.jar`
   * `tachyon-client-0.5.0.jar`

Commit:
Bypass SerializationDebugger for now (SPARK-7180)

SparkQA commented:
[Test build #31109 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31109/consoleFull) for   PR 5685 at commit [`9419efe`](https://github.com/apache/spark/commit/9419efea98e190766887a9ab5fbac57f45bf007c).

andrewor14 commented:
I found the issue. Turns out `SerializationDebugger` currently cannot serialize `ClosureCleanerSuite2` due to [SPARK-7180](https://issues.apache.org/jira/browse/SPARK-7180). In a nutshell, it turns out that the debugger currently doesn't handle the case where an object both (1) inherits a serializable parent, and (2) has a non-serializable field. The reason why it passed for me locally before was because the debugger was somehow not [enabled](https://github.com/apache/spark/blob/4d9e560b5470029143926827b1cb9d72a0bfbeff/core/src/main/scala/org/apache/spark/serializer/SerializationDebugger.scala#L62), and I suspect this has something to do with the fact that I'm running Java 6. I was able to reproduce the exception locally by always enabling the debugger.

Anyway, I have disabled the debugger in the test for now and hopefully it will pass.

SparkQA commented:
[Test build #31109 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31109/consoleFull) for   PR 5685 at commit [`9419efe`](https://github.com/apache/spark/commit/9419efea98e190766887a9ab5fbac57f45bf007c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `   *   class SomethingNotSerializable `
  * `      logDebug(s" + cloning the object $obj of class $`

 * This patch does not change any dependencies.

Commit:
Revert "Bypass SerializationDebugger for now (SPARK-7180)"

This reverts commit 9419efea98e190766887a9ab5fbac57f45bf007c.

Commit:
Merge branch 'master' of github.com:apache/spark into closure-cleaner

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

andrewor14 commented:
I've reverted the previous commit since https://github.com/apache/spark/commit/bf35edd9d4b8b11df9f47b6ff43831bc95f06322 was merged into master and the workaround for SPARK-7180 is no longer needed.

SparkQA commented:
[Test build #31172 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31172/consoleFull) for   PR 5685 at commit [`9187066`](https://github.com/apache/spark/commit/9187066ef9834d62d488b0ccc3be70eea514ae2c).

SparkQA commented:
[Test build #31172 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31172/consoleFull) for   PR 5685 at commit [`9187066`](https://github.com/apache/spark/commit/9187066ef9834d62d488b0ccc3be70eea514ae2c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31172/
Test PASSed.

Commit:
In sc.runJob, actually clean the inner closure

If a closure is passed into another closure as a reference, the
inner closure will be a field of the outer closure. In sc.runJob,
we used to only clean the outer closure, leaving the inner one
uncleaned.

Simple reproduction:

Wrap RDD#take in a closure. For instance:

  // The body "..." here contains a return statement
  def take(num: Int): Array[T] = (1 to 1).foreach { _ => ... }

Now if you call `sc.parallelize(1 to 10).take(5)`, the closure
cleaner will not be able to find the return statement in the
`foreach` closure. This is because it's not even cleaning the
`foreach` closure. Instead, it will fail with a not serializable
exception complaining the internal java.lang.Object
$nonLocalReturnKey is not serializable.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31200 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31200/consoleFull) for   PR 5685 at commit [`3998168`](https://github.com/apache/spark/commit/399816819e0f0bc9fde9fac1a3d6b6f6ef641a74).

SparkQA commented:
[Test build #31200 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31200/consoleFull) for   PR 5685 at commit [`3998168`](https://github.com/apache/spark/commit/399816819e0f0bc9fde9fac1a3d6b6f6ef641a74).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `   *   class SomethingNotSerializable `
  * `      logDebug(s" + cloning the object $obj of class $`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31200/
Test FAILed.

Commit:
Guard against NPE if CC is used outside of an application

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31217 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31217/consoleFull) for   PR 5685 at commit [`e909a42`](https://github.com/apache/spark/commit/e909a4222a132b1fc4fff87a111ece7d32b82765).

SparkQA commented:
[Test build #31217 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31217/consoleFull) for   PR 5685 at commit [`e909a42`](https://github.com/apache/spark/commit/e909a4222a132b1fc4fff87a111ece7d32b82765).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `   *   class SomethingNotSerializable `
  * `      logDebug(s" + cloning the object $obj of class $`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31217/
Test PASSed.

zsxwing commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1627,7 +1627,10 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
-    runJob(rdd, (context: TaskContext, iter: Iterator[T]) => func(iter), partitions, allowLocal)
+    // We must clean `func` here before using it in another closure below
+    // Otherwise, the closure cleaner will only clean the outer closure but not `func`
+    val cleanedFunc = clean(func)
```
Do we need to apply such fix to other places? E.g., RDD.scala

```Scala
  def mapPartitions[U: ClassTag](
      f: Iterator[T] => Iterator[U], preservesPartitioning: Boolean = false): RDD[U] = {
    val func = (context: TaskContext, index: Int, iter: Iterator[T]) => f(iter)
    new MapPartitionsRDD(this, sc.clean(func), preservesPartitioning)
  }
```

andrewor14 replied:
Yeah, I think so. This seems like a separate issue so maybe I should file another JIRA for it.

andrewor14 replied:
Ok, I have decided to fix this in a separate patch since it's really a separate issue. I have filed [SPARK-7237](https://issues.apache.org/jira/browse/SPARK-7237) for that

Commit:
Revert "Guard against NPE if CC is used outside of an application"

This reverts commit e909a4222a132b1fc4fff87a111ece7d32b82765.

Commit:
Revert "In sc.runJob, actually clean the inner closure"

This reverts commit 399816819e0f0bc9fde9fac1a3d6b6f6ef641a74.

andrewor14 commented:
I have moved the last two commits into a separate patch #5787 (SPARK-7237). It is really a distinct issue  that can be merged without this patch and should not further inflate the size of this one.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31315 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31315/consoleFull) for   PR 5685 at commit [`26c7aba`](https://github.com/apache/spark/commit/26c7abaca5ea6a76f6839c29d9c209e9aaee4948).

SparkQA commented:
[Test build #31315 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31315/consoleFull) for   PR 5685 at commit [`26c7aba`](https://github.com/apache/spark/commit/26c7abaca5ea6a76f6839c29d9c209e9aaee4948).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `   *   class SomethingNotSerializable `
  * `      logDebug(s" + cloning the object $obj of class $`

 * This patch does not change any dependencies.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31315/
Test PASSed.

pwendell commented on core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala:
```diff
@@ -101,21 +115,124 @@ private[spark] object ClosureCleaner extends Logging {
+   *       def y = 2
+   *       scope("two") { println(y + 1) }
+   *     }
+   *     def scope(name: String)(body: => Unit) = body
```
maybe move this def up to define it before you use it

pwendell commented on core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala:
```diff
@@ -101,21 +115,124 @@ private[spark] object ClosureCleaner extends Logging {
+    logDebug(s"+++ Cleaning closure $func (${func.getClass.getName}}) +++")
+
+    // A list of classes that represents closures enclosed in the given one
     val innerClasses = getInnerClasses(func)
```
should this be called `innerClosureClasses` and `getInnerClosureClasses`?

pwendell commented on core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala:
```diff
@@ -55,10 +58,14 @@ private[spark] object ClosureCleaner extends Logging {
   private def getOuterClasses(obj: AnyRef): List[Class[_]] = {
```
Should we merge this and `getOuterObjects` into one function that just returns pairs of objects and their classes? Maybe not just to keep the changes minimal?

andrewor14 replied:
let's do that later

pwendell commented on core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala:
```diff
@@ -184,19 +335,17 @@ private[spark] object ClosureCleaner extends Logging {
+private class ReturnStatementFinder extends ClassVisitor(ASM4) {
   override def visitMethod(access: Int, name: String, desc: String,
       sig: String, exceptions: Array[String]): MethodVisitor = {
     if (name.contains("apply")) {
```
We can do this later, but if possible we should try to return the line number of the return statement if possible (is that even possible?).

andrewor14 replied:
let's do that later

pwendell commented on core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala:
```diff
@@ -126,34 +243,66 @@ private[spark] object ClosureCleaner extends Logging {
+      // the already populated accessed fields map of the starting closure
+      if (cleanTransitively && isClosure(clone.getClass)) {
+        logDebug(s" + cleaning cloned closure $clone recursively (${cls.getName})")
+        clean(clone, checkSerializable, cleanTransitively, accessedFields)
```
should `checkSerailizable` be false here? It seems like theoretically we only need to check the very final closure. It's a bit hard for me to reason about whether we could have a weird state where `checkSerializable` throws false positive errors during recursion.

pwendell commented on core/src/test/scala/org/apache/spark/util/ClosureCleanerSuite2.scala:
```diff
@@ -0,0 +1,562 @@
+   * Helper method for testing whether closure cleaning works as expected.
+   * This cleans the given closure twice, with and without transitive cleaning.
+   */
+  private def testClean(
```
Maybe these should be called `verifyCleaning` or something. Because they are really verifying behavior according to what you pass as `serializableBefore`/`serializableAfter`. Also it would be good to document those parameters... it's not super clear what they mean.

andrewor14 replied:
ok, I thought it would be clear from the name but apparently not

pwendell commented on core/src/test/scala/org/apache/spark/util/ClosureCleanerSuite2.scala:
```diff
@@ -0,0 +1,562 @@
+      val fields3 = findAccessedFields(closure3, outerClasses3, findTransitively = false)
+      val fields4 = findAccessedFields(closure4, outerClasses4, findTransitively = false)
+      assert(fields1.isEmpty)
+      // "test1" < "FunSuite#test" < ClosureCleanerSuite2
```
I think the middle one should maybe be clear that it's really the anonymous class passed into the `test` method, right?

pwendell commented:
Hey Andrew,

This is looking good. The code is quite dense but as far as I can tell, this is correct. I left some more surface level comments, if you can get to those I think it will be good to go.

zsxwing commented on core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala:
```diff
@@ -55,10 +58,14 @@ private[spark] object ClosureCleaner extends Logging {
+      // The outer pointer may be null if we have cleaned this closure before
+      if (outer != null) {
+        if (isClosure(f.getType)) {
+          return f.getType :: getOuterClasses(f.get(obj))
```
"f.get(obj)"  -> "outer"

andrewor14 replied:
oops yes, thanks.

zsxwing commented on core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala:
```diff
@@ -68,15 +75,22 @@ private[spark] object ClosureCleaner extends Logging {
+      // The outer pointer may be null if we have cleaned this closure before
+      if (outer != null) {
+        if (isClosure(f.getType)) {
+          return f.get(obj) :: getOuterObjects(f.get(obj))
```
"f.get(obj)" -> "outer"

zsxwing commented on core/src/main/scala/org/apache/spark/util/ClosureCleaner.scala:
```diff
@@ -68,15 +75,22 @@ private[spark] object ClosureCleaner extends Logging {
+        if (isClosure(f.getType)) {
+          return f.get(obj) :: getOuterObjects(f.get(obj))
+        } else {
+          return f.get(obj) :: Nil // Stop at the first $outer that is not a closure
```
"f.get(obj)" -> "outer"

pwendell commented:
@andrewor14 if you can update this I think this one is good to go.

Commit:
Merge branch 'master' of github.com:apache/spark into closure-cleaner

Commit:
Address comments

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31637 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31637/consoleFull) for   PR 5685 at commit [`26c5072`](https://github.com/apache/spark/commit/26c507257a80273d5941e3c974c15d4721cc3b4b).

SparkQA commented:
[Test build #31637 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31637/consoleFull) for   PR 5685 at commit [`26c5072`](https://github.com/apache/spark/commit/26c507257a80273d5941e3c974c15d4721cc3b4b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `   *   class SomethingNotSerializable `
  * `      logDebug(s" + cloning the object $obj of class $`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31637/
Test FAILed.

Commit:
Fix tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31639 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31639/consoleFull) for   PR 5685 at commit [`ea874bc`](https://github.com/apache/spark/commit/ea874bc6c9cdede04cf4bd68b0053811cdef11a9).

SparkQA commented:
[Test build #31639 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31639/consoleFull) for   PR 5685 at commit [`ea874bc`](https://github.com/apache/spark/commit/ea874bc6c9cdede04cf4bd68b0053811cdef11a9).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `   *   class SomethingNotSerializable `
  * `      logDebug(s" + cloning the object $obj of class $`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31639/
Test FAILed.

Commit:
Fix style

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31642 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31642/consoleFull) for   PR 5685 at commit [`0bbe77f`](https://github.com/apache/spark/commit/0bbe77f10a23e8fe96d587b69266c9cbc85b9ed7).

andrewor14 commented:
Ok the latest changes are G2G from my side.

SparkQA commented:
**[Test build #31642 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31642/consoleFull)**     for PR 5685 at commit [`0bbe77f`](https://github.com/apache/spark/commit/0bbe77f10a23e8fe96d587b69266c9cbc85b9ed7)     after a configured wait of `150m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31642/
Test FAILed.

Commit:
Revert a small change that affected streaming

DStream#transform isn't cleaning closures correctly. It is passing
an RDD to ClosureCleaner#clean. We should fix this separately
outside of this patch.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31654 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31654/consoleFull) for   PR 5685 at commit [`cd46230`](https://github.com/apache/spark/commit/cd4623006d0d30c1fcd66eb7c947eab4d201e43b).

andrewor14 commented:
It turns out that in streaming we currently pass an RDD into `ClosureCleaner#clean` so we can't do the type safety check that you supported @pwendell. I will fix this separately later in SPARK-7318.

pwendell commented:
@andrewor14 okay makes sense. Thanks for all the hard work on this andrew LGTM pending tests.

SparkQA commented:
[Test build #31654 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31654/consoleFull) for   PR 5685 at commit [`cd46230`](https://github.com/apache/spark/commit/cd4623006d0d30c1fcd66eb7c947eab4d201e43b).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `   *   class SomethingNotSerializable `
  * `      logDebug(s" + cloning the object $obj of class $`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31654/
Test PASSed.
